Crash when there are no more repositories in github provider

`rovers` keeps crashing after it has consumed all repositories from github provider:

```
t=2018-04-03T16:52:56+0000 lvl=info msg="No providers added using --provider option. Executing all known providers" providers="[github cgit bitbucket]" caller=cmd_repo_providers.go:32
t=2018-04-03T16:52:56+0000 lvl=info msg="Creating github provider" caller=cmd_repo_providers.go:42
t=2018-04-03T16:52:56+0000 lvl=info msg="Creating cgit provider" caller=cmd_repo_providers.go:49
t=2018-04-03T16:52:56+0000 lvl=info msg="Creating bitbucket provider" caller=cmd_repo_providers.go:56
t=2018-04-03T16:52:56+0000 lvl=info msg=Watcher time=1h0m0s caller=cmd_repo_providers.go:65
t=2018-04-03T16:52:57+0000 lvl=info msg="checkpoint empty, trying to get checkpoint" caller=provider.go:43
t=2018-04-03T16:52:57+0000 lvl=info msg="obtaining page for Bing search" query="\"powered by cgit\"||\"generated by cgit\"||\"Commits per author per week\"" offset=0 api URL="https://api.cognitive.microsoft.com/bing/v5.0/search?count=50&offset=0&q=%22powered+by+cgit%22%7C%7C%22generated+by+cgit%22%7C%7C%22Commits+per+author+per+week%22&responseFilter=Webpages" caller=bing.go:85
t=2018-04-03T16:52:57+0000 lvl=eror msg="error getting cgit urls from Searcher" error="Invalid key" caller=provider.go:133
t=2018-04-03T16:52:57+0000 lvl=warn msg="no scrapers found, sending an EOF because we have no data" caller=provider.go:169
t=2018-04-03T16:52:57+0000 lvl=info msg="no more repositories, waiting for more..." time to sleep=1h0m0s caller=watcher.go:62
t=2018-04-03T16:52:57+0000 lvl=info msg="checkpoint found" checkpoint=2018-04-03T16:42:15.626924+00:00 caller=provider.go:56
t=2018-04-03T16:53:01+0000 lvl=info msg="no more repositories, waiting for more..." time to sleep=1h0m0s caller=watcher.go:62
t=2018-04-03T16:53:07+0000 lvl=info msg="no repositories into cache, getting more repositories" checkpoint=124333782 caller=provider.go:51
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb4f65f]

goroutine 35 [running]:
github.com/src-d/rovers/providers/github.(*provider).saveRepos.func1(0xc4201f2020, 0xc4201f2020, 0xc42041cd10)
	/home/travis/gopath/src/github.com/src-d/rovers/providers/github/provider.go:124 +0x5f
github.com/src-d/rovers/providers/github/model.(*RepositoryStore).Transaction.func1(0xc420372750, 0xc420372750, 0x0)
	/home/travis/gopath/src/github.com/src-d/rovers/providers/github/model/kallax.go:244 +0x64
github.com/src-d/rovers/vendor/gopkg.in/src-d/go-kallax%2ev1.(*Store).Transaction(0xc42019eed0, 0xc420649a50, 0x8d, 0x131f)
	/home/travis/gopath/src/github.com/src-d/rovers/vendor/gopkg.in/src-d/go-kallax.v1/store.go:435 +0x10f
github.com/src-d/rovers/providers/github/model.(*RepositoryStore).Transaction(0xc42000c270, 0xc420649a90, 0xbf1300, 0xc4203fe2b0)
	/home/travis/gopath/src/github.com/src-d/rovers/providers/github/model/kallax.go:245 +0x60
github.com/src-d/rovers/providers/github.(*provider).saveRepos(0xc420116f40, 0xc420390d80, 0x64, 0x8d, 0x0, 0x2)
	/home/travis/gopath/src/github.com/src-d/rovers/providers/github/provider.go:132 +0x81
github.com/src-d/rovers/providers/github.(*provider).requestNextPage(0xc420116f40, 0x7692ed6, 0xc42024fbf0, 0x2, 0x2, 0x0, 0x0)
	/home/travis/gopath/src/github.com/src-d/rovers/providers/github/provider.go:99 +0x8a
github.com/src-d/rovers/providers/github.(*provider).Next(0xc420116f40, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/src-d/rovers/providers/github/provider.go:52 +0x1cd
github.com/src-d/rovers/core.(*Watcher).handleProviderResult(0xc4201f0150, 0x11af420, 0xc420116f40, 0x180001, 0x0)
	/home/travis/gopath/src/github.com/src-d/rovers/core/watcher.go:57 +0x4c
github.com/src-d/rovers/core.(*Watcher).Start.func1(0xc4201f0150, 0x11af420, 0xc420116f40)
	/home/travis/gopath/src/github.com/src-d/rovers/core/watcher.go:48 +0x43
created by github.com/src-d/rovers/core.(*Watcher).Start
	/home/travis/gopath/src/github.com/src-d/rovers/core/watcher.go:52 +0x71
```

It seems that the `repositories` slice contains a nil in `providers/github/provider.go#saveRepos` but I have not been able to create a json that decodes to a `[ nil ]`.

```go
func (gp *provider) saveRepos(repositories []*model.Repository) error {
	return gp.repositoriesStore.Transaction(func(s *model.RepositoryStore) error {
		for _, repo := range repositories {
			repo.ID = kallax.NewULID()
			err := s.Insert(repo)
			if err != nil {
				return err
			}
		}

		return nil
	})
}
```

We can skip non valid repo pointers but it would be nice to understand why this is happening.

--- a/providers/github/client.go
+++ b/providers/github/client.go
@@ -30,7 +30,8 @@ type response struct {
 }
 
 type client struct {
-	c *http.Client
+	c        *http.Client
+	endpoint string
 }
 
 func newClient(token string) *client {
@@ -43,7 +44,7 @@ func newClient(token string) *client {
 
 	c.Timeout = httpTimeout
 
-	return &client{c}
+	return &client{c, githubApiURL}
 }
 
 // Repositories returns a response with the next page id and a list of Repositories.
@@ -51,7 +52,7 @@ func newClient(token string) *client {
 func (c *client) Repositories(since int) (*response, error) {
 	start := time.Now()
 
-	u := fmt.Sprintf(githubApiURL, since)
+	u := fmt.Sprintf(c.endpoint, since)
 	res, err := c.c.Get(u)
 	if err != nil {
 		return nil, err
@@ -81,7 +82,6 @@ func (c *client) Repositories(since int) (*response, error) {
 
 	total := c.toInt(res.Header.Get(rateLimitLimitHeader))
 	remaining := c.toInt(res.Header.Get(rateLimitRemainingHeader))
-
 	minRequestDuration := time.Hour / time.Duration(total)
 	defer func() {
 		needsWait := minRequestDuration - time.Since(start)

--- a/providers/github/client.go
+++ b/providers/github/client.go
@@ -3,11 +3,12 @@ package github
 import (
 	"encoding/json"
 	"fmt"
-	"golang.org/x/oauth2"
 	"io"
 	"net/http"
 	"strconv"
 	"time"
+
+	"golang.org/x/oauth2"
 
 	"github.com/src-d/rovers/providers/github/model"
 )
@@ -62,6 +63,21 @@ func (c *client) Repositories(since int) (*response, error) {
 	}
 
 	repositories, err := c.decode(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// remove those repositories that GitHub API encoded as null
+	// in the JSON reponse and were decoded as a nil element in the
+	// *model.Repository slice.
+	repos := make([]*model.Repository, 0, len(repositories))
+	for _, repo := range repositories {
+		if repo != nil {
+			repos = append(repos, repo)
+		}
+	}
+
+	repositories = repos
 
 	total := c.toInt(res.Header.Get(rateLimitLimitHeader))
 	remaining := c.toInt(res.Header.Get(rateLimitRemainingHeader))

--- a/providers/github/provider_test.go
+++ b/providers/github/provider_test.go
@@ -4,6 +4,8 @@ import (
 	"database/sql"
 	"errors"
 	"io"
+	"net/http"
+	"net/http/httptest"
 	"testing"
 
 	"github.com/src-d/rovers/core"
@@ -113,4 +115,149 @@ func (s *GithubProviderSuite) TestGithubProvider_Next_Retry(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(repoUrl, NotNil)
 	c.Assert(repoUrl, DeepEquals, repoUrl2)
+}
+
+func (s *GithubProviderSuite) TestGithubProviderNilRepositories(c *C) {
+	const response string = `
+[
+    null,
+    {
+	"id": 40908251,
+	"name": "rovers",
+	"full_name": "src-d/rovers",
+	"owner": {
+	  "login": "src-d",
+	  "id": 15128793,
+	  "avatar_url": "https://avatars2.githubusercontent.com/u/15128793?v=4",
+	  "gravatar_id": "",
+	  "url": "https://api.github.com/users/src-d",
+	  "html_url": "https://github.com/src-d",
+	  "followers_url": "https://api.github.com/users/src-d/followers",
+	  "following_url": "https://api.github.com/users/src-d/following{/other_user}",
+	  "gists_url": "https://api.github.com/users/src-d/gists{/gist_id}",
+	  "starred_url": "https://api.github.com/users/src-d/starred{/owner}{/repo}",
+	  "subscriptions_url": "https://api.github.com/users/src-d/subscriptions",
+	  "organizations_url": "https://api.github.com/users/src-d/orgs",
+	  "repos_url": "https://api.github.com/users/src-d/repos",
+	  "events_url": "https://api.github.com/users/src-d/events{/privacy}",
+	  "received_events_url": "https://api.github.com/users/src-d/received_events",
+	  "type": "Organization",
+	  "site_admin": false
+	},
+	"private": false,
+	"html_url": "https://github.com/src-d/rovers",
+	"description": "Rovers is a service to retrieve repository URLs from multiple repository hosting providers.",
+	"fork": false,
+	"url": "https://api.github.com/repos/src-d/rovers",
+	"forks_url": "https://api.github.com/repos/src-d/rovers/forks",
+	"keys_url": "https://api.github.com/repos/src-d/rovers/keys{/key_id}",
+	"collaborators_url": "https://api.github.com/repos/src-d/rovers/collaborators{/collaborator}",
+	"teams_url": "https://api.github.com/repos/src-d/rovers/teams",
+	"hooks_url": "https://api.github.com/repos/src-d/rovers/hooks",
+	"issue_events_url": "https://api.github.com/repos/src-d/rovers/issues/events{/number}",
+	"events_url": "https://api.github.com/repos/src-d/rovers/events",
+	"assignees_url": "https://api.github.com/repos/src-d/rovers/assignees{/user}",
+	"branches_url": "https://api.github.com/repos/src-d/rovers/branches{/branch}",
+	"tags_url": "https://api.github.com/repos/src-d/rovers/tags",
+	"blobs_url": "https://api.github.com/repos/src-d/rovers/git/blobs{/sha}",
+	"git_tags_url": "https://api.github.com/repos/src-d/rovers/git/tags{/sha}",
+	"git_refs_url": "https://api.github.com/repos/src-d/rovers/git/refs{/sha}",
+	"trees_url": "https://api.github.com/repos/src-d/rovers/git/trees{/sha}",
+	"statuses_url": "https://api.github.com/repos/src-d/rovers/statuses/{sha}",
+	"languages_url": "https://api.github.com/repos/src-d/rovers/languages",
+	"stargazers_url": "https://api.github.com/repos/src-d/rovers/stargazers",
+	"contributors_url": "https://api.github.com/repos/src-d/rovers/contributors",
+	"subscribers_url": "https://api.github.com/repos/src-d/rovers/subscribers",
+	"subscription_url": "https://api.github.com/repos/src-d/rovers/subscription",
+	"commits_url": "https://api.github.com/repos/src-d/rovers/commits{/sha}",
+	"git_commits_url": "https://api.github.com/repos/src-d/rovers/git/commits{/sha}",
+	"comments_url": "https://api.github.com/repos/src-d/rovers/comments{/number}",
+	"issue_comment_url": "https://api.github.com/repos/src-d/rovers/issues/comments{/number}",
+	"contents_url": "https://api.github.com/repos/src-d/rovers/contents/{+path}",
+	"compare_url": "https://api.github.com/repos/src-d/rovers/compare/{base}...{head}",
+	"merges_url": "https://api.github.com/repos/src-d/rovers/merges",
+	"archive_url": "https://api.github.com/repos/src-d/rovers/{archive_format}{/ref}",
+	"downloads_url": "https://api.github.com/repos/src-d/rovers/downloads",
+	"issues_url": "https://api.github.com/repos/src-d/rovers/issues{/number}",
+	"pulls_url": "https://api.github.com/repos/src-d/rovers/pulls{/number}",
+	"milestones_url": "https://api.github.com/repos/src-d/rovers/milestones{/number}",
+	"notifications_url": "https://api.github.com/repos/src-d/rovers/notifications{?since,all,participating}",
+	"labels_url": "https://api.github.com/repos/src-d/rovers/labels{/name}",
+	"releases_url": "https://api.github.com/repos/src-d/rovers/releases{/id}",
+	"deployments_url": "https://api.github.com/repos/src-d/rovers/deployments",
+	"created_at": "2015-08-17T15:59:59Z",
+	"updated_at": "2018-03-22T08:44:49Z",
+	"pushed_at": "2018-04-19T09:11:13Z",
+	"git_url": "git://github.com/src-d/rovers.git",
+	"ssh_url": "git@github.com:src-d/rovers.git",
+	"clone_url": "https://github.com/src-d/rovers.git",
+	"svn_url": "https://github.com/src-d/rovers",
+	"homepage": "",
+	"size": 4171,
+	"stargazers_count": 4,
+	"watchers_count": 4,
+	"language": "HTML",
+	"has_issues": true,
+	"has_projects": false,
+	"has_downloads": true,
+	"has_wiki": true,
+	"has_pages": false,
+	"forks_count": 9,
+	"mirror_url": null,
+	"archived": false,
+	"open_issues_count": 8,
+	"license": {
+	  "key": "gpl-3.0",
+	  "name": "GNU General Public License v3.0",
+	  "spdx_id": "GPL-3.0",
+	  "url": "https://api.github.com/licenses/gpl-3.0"
+	},
+	"forks": 9,
+	"open_issues": 8,
+	"watchers": 4,
+	"default_branch": "master",
+	"organization": {
+	  "login": "src-d",
+	  "id": 15128793,
+	  "avatar_url": "https://avatars2.githubusercontent.com/u/15128793?v=4",
+	  "gravatar_id": "",
+	  "url": "https://api.github.com/users/src-d",
+	  "html_url": "https://github.com/src-d",
+	  "followers_url": "https://api.github.com/users/src-d/followers",
+	  "following_url": "https://api.github.com/users/src-d/following{/other_user}",
+	  "gists_url": "https://api.github.com/users/src-d/gists{/gist_id}",
+	  "starred_url": "https://api.github.com/users/src-d/starred{/owner}{/repo}",
+	  "subscriptions_url": "https://api.github.com/users/src-d/subscriptions",
+	  "organizations_url": "https://api.github.com/users/src-d/orgs",
+	  "repos_url": "https://api.github.com/users/src-d/repos",
+	  "events_url": "https://api.github.com/users/src-d/events{/privacy}",
+	  "received_events_url": "https://api.github.com/users/src-d/received_events",
+	  "type": "Organization",
+	  "site_admin": false
+	},
+	"network_count": 9,
+	"subscribers_count": 15
+      },
+      null
+]
+`
+
+	mockedServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set(rateLimitLimitHeader, "123456789")
+		w.Header().Set(rateLimitRemainingHeader, "0")
+		w.Write([]byte(response))
+	}))
+	defer mockedServer.Close()
+
+	s.provider.(*provider).apiClient.endpoint = mockedServer.URL + "?%d=foo"
+	defer func() {
+		s.provider.(*provider).apiClient.endpoint = githubApiURL
+	}()
+
+	mention, err := s.provider.Next()
+	c.Assert(err, IsNil)
+	c.Assert(mention, NotNil)
+	c.Assert(mention.Endpoint, Equals, "git://github.com/src-d/rovers")
+	err = s.provider.Ack(nil)
+	c.Assert(err, IsNil)
 }

